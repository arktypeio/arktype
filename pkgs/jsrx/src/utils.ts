import { SyncOptions } from "execa"
import { command } from "@re-do/utils/dist/command"

export const run = command
export const $ = (cmd: string, options?: SyncOptions) => () => run(cmd, options)
