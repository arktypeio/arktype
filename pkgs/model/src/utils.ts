import { transform, TypeCategory, isRecursible } from "@re-/tools"
import { ExtractableDefinition } from "./internal.js"

export const typeOf = (value: any): ExtractableDefinition => {
    const typeMap: { [Category in TypeCategory]: () => ExtractableDefinition } =
        {
            string: () => `'${value}'`,
            number: () => value,
            bigint: () => value,
            object: () =>
                value === null
                    ? "null"
                    : transform(value, ([k, v]) => [k, typeOf(v)]),
            undefined: () => "undefined",
            symbol: () => "symbol",
            function: () => "function",
            boolean: () => (value ? "true" : "false")
        }
    return typeMap[typeof value]()
}
