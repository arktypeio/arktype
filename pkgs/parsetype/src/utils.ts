// As of TS 4.5.0-beta, this is required to prevent TS2742 in multifile.test.ts
export * from "@re-do/utils"
