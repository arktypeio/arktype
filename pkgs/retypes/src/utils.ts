// As of TS 4.5.0-beta, this is required to prevent TS2742 in multifile.test.ts
export * from "@re-do/utils"
export * from "./main.js"
export * from "./common.js"
