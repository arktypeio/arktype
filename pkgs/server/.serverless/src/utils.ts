export const ifExists = async <T>(find: () => Promise<T>) => {
    try {
        return await find()
    } catch {
        return null
    }
}

export const APP_SECRET = process.env.APP_SECRET as string
