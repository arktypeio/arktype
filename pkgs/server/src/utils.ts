export const APP_SECRET = process.env.APP_SECRET as string
