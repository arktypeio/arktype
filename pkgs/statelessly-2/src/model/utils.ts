import {
    Object as ToolbeltObject,
    Union as ToolbeltUnion,
    List as ToolbeltList,
    Any as ToolbeltAny,
    Function as ToolbeltFunction
} from "ts-toolbelt"

import {
    NonRecursible,
    Unlisted,
    KeyValuate,
    ExcludeCyclic,
    Join,
    NonObject
} from "@re-do/utils"

export type TypeError<Description extends string> = Description

export type ForceEvaluate<T, Deep extends boolean = true> = ToolbeltAny.Compute<
    T,
    Deep extends true ? "deep" : "flat"
>

export type Cast<A, B> = A extends B ? A : B

type NarrowRecurse<T> =
    | (T extends [] ? [] : never)
    | (T extends Narrowable ? T : never)
    | {
          [K in keyof T]: T[K] extends Function ? T[K] : Narrow<T[K]>
      }

type NarrowRoot<T, Narrowed = NarrowRecurse<T>> = Narrowed | Cast<T, Narrowed>

type Narrowable = string | number | bigint | boolean

export type Narrow<T> = NarrowRoot<T>

const narrow = <T>(t: Narrow<T>): T => [] as any as T

const f = narrow({
    z: (x: [1, "f", true, [3, "af"]]) => [true, false, ["a", 5]]
})

type ListPossibleTypesRecurse<
    U,
    LN extends any[] = [],
    LastU = ToolbeltUnion.Last<U>
> = {
    0: ListPossibleTypesRecurse<
        Exclude<U, LastU>,
        ToolbeltList.Prepend<LN, LastU>
    >
    1: LN
}[[U] extends [never] ? 1 : 0]

export type ListPossibleTypes<U> = ListPossibleTypesRecurse<U> extends infer X
    ? Cast<X, any[]>
    : never

export type StringifyPossibleTypes<U extends string | number> = Join<
    ListPossibleTypes<U>,
    ", "
>

type ExactFunction<F extends Function, ExpectedType> = F extends (
    ...args: infer Args
) => infer Return
    ? ExpectedType extends (...args: infer ExpectedArgs) => infer ExpectedReturn
        ? (...args: Exact<Args, ExpectedArgs>) => Exact<Return, ExpectedReturn>
        : never
    : never

export type Exact<T, ExpectedType> = ExpectedType extends unknown
    ? T extends ExpectedType
        ? T extends NonObject
            ? T
            : T extends Function
            ? any // ExactFunction<T, ExpectedType>
            : {
                  [K in keyof T]: K extends keyof ExpectedType
                      ? Exact<T[K], ExpectedType[K]>
                      : TypeError<`Invalid property '${K &
                            (
                                | string
                                | number
                            )}'. Valid properties are: ${StringifyPossibleTypes<
                            keyof ExpectedType & (string | number)
                        >}`>
              }
        : ExpectedType
    : never

const exact = <T>(t: Exact<T, () => { a: 5 }>) => {}

exact(() => ({ a: 5 }))
