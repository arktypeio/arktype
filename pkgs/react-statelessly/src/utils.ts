import { Store as BaseStore, Actions } from "statelessly"
import { ValueFrom } from "@re-do/utils"

export type Store<T extends object, A extends Actions<T>> = BaseStore<T, A> & {
    useQuery: ValueFrom<BaseStore<T, A>, "query">
    useGet: ValueFrom<BaseStore<T, A>, "get">
    useGetState: ValueFrom<BaseStore<T, A>, "getState">
}
