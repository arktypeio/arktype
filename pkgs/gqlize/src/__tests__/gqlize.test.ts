import { join } from "path"
import { fromEntries, camelCase } from "@re-do/utils"
import { gqlize, getObjectDefinition } from ".."

describe("gqlize", () => {
    const schema = join(__dirname, "schema.gql")
    it("works with default options", async () => {
        expect(gqlize({ schema })).toMatchSnapshot("defaults")
    })
    it("can generate new queries from defaults", async () => {
        gqlize({
            schema,
            queries: {
                me: {
                    branch: (data, schema) => {
                        const userType = getObjectDefinition("User", schema)
                        return fromEntries(
                            userType.fields.map((field) => [
                                camelCase(["my", field.name.value]),
                                {
                                    ...data,
                                    fields: data.output.fields.filter(
                                        (resultField) =>
                                            resultField.name.value ===
                                            field.name.value
                                    )
                                }
                            ])
                        )
                    }
                }
            }
        })
    }),
        it("can transform existing queries", async () => {
            const transformed = gqlize({
                schema,
                queries: {
                    me: {
                        map: (data) => ({ ...data, name: "myself" })
                    }
                }
            })
            expect(transformed).toContain("query myself")
            expect(transformed).not.toContain("query me")
        }),
        it("can globally transform queries", async () => {
            const transformed = gqlize({
                schema,
                transformOutputs: (fields) =>
                    fields.filter((field) => field.name.value !== "user")
            })
            expect(transformed).not.toContain("user")
        })
})
