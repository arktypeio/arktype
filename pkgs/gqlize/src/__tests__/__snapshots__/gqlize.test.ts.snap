// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gqlize works with default options: defaults 1`] = `
"query me {
  me {
    email
    first
    id
    last
    password
    steps {
      data {
        assertText {
          expected
          id
          selector
        }
        assertVisibility {
          expected
          id
          selector
        }
        click {
          double
          id
          selector
        }
        go {
          id
          url
        }
        hover {
          duration
          id
          selector
        }
        id
        key {
          id
          key
        }
        set {
          id
          selector
          value
        }
      }
      id
      kind
      tests {
        id
        name
        tags {
          id
          name
        }
      }
    }
    tags {
      id
      name
      test {
        id
        name
        steps {
          data {
            id
          }
          id
          kind
        }
      }
    }
    tests {
      id
      name
      steps {
        data {
          assertText {
            expected
            id
            selector
          }
          assertVisibility {
            expected
            id
            selector
          }
          click {
            double
            id
            selector
          }
          go {
            id
            url
          }
          hover {
            duration
            id
            selector
          }
          id
          key {
            id
            key
          }
          set {
            id
            selector
            value
          }
        }
        id
        kind
      }
      tags {
        id
        name
      }
    }
  }
}
mutation createTest(
  $name: String!
  $steps: StepCreateWithoutTestsInput
  $tags: TagCreateWithoutTestInput
) {
  createTest(data: { name: $name, steps: $steps, tags: $tags }) {
    id
    name
    steps {
      data {
        assertText {
          expected
          id
          selector
        }
        assertVisibility {
          expected
          id
          selector
        }
        click {
          double
          id
          selector
        }
        go {
          id
          url
        }
        hover {
          duration
          id
          selector
        }
        id
        key {
          id
          key
        }
        set {
          id
          selector
          value
        }
      }
      id
      kind
      user {
        email
        first
        id
        last
        password
        tags {
          id
          name
        }
      }
    }
    tags {
      id
      name
      user {
        email
        first
        id
        last
        password
        steps {
          data {
            id
          }
          id
          kind
        }
      }
    }
    user {
      email
      first
      id
      last
      password
      steps {
        data {
          assertText {
            expected
            id
            selector
          }
          assertVisibility {
            expected
            id
            selector
          }
          click {
            double
            id
            selector
          }
          go {
            id
            url
          }
          hover {
            duration
            id
            selector
          }
          id
          key {
            id
            key
          }
          set {
            id
            selector
            value
          }
        }
        id
        kind
      }
      tags {
        id
        name
      }
    }
  }
}
mutation signIn($email: String!, $password: String!) {
  signIn(data: { email: $email, password: $password })
}
mutation signUp(
  $email: String!
  $first: String!
  $last: String!
  $password: String!
) {
  signUp(
    data: { email: $email, first: $first, last: $last, password: $password }
  )
}
"
`;
